aws/src/main/java/org/apache/iceberg/aws/S3FileIOAwsClientFactories.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.aws;
+
+import java.util.Map;
+import org.apache.iceberg.aws.s3.S3FileIOAwsClientFactory;
+import org.apache.iceberg.aws.s3.S3FileIOProperties;
+import org.apache.iceberg.common.DynConstructors;
+import org.apache.iceberg.relocated.com.google.common.base.Strings;
+import org.apache.iceberg.util.PropertyUtil;
+
+public class S3FileIOAwsClientFactories {
+
+  private S3FileIOAwsClientFactories() {}
+
+  /**
+   * Attempts to load an AWS client factory class for S3 file IO defined in the catalog property
+   * {@link S3FileIOProperties#CLIENT_FACTORY}. If the property wasn't set, fallback to {@link
+   * AwsClientFactories#from(Map) to intialize an AWS client factory class}
+   *
+   * @param properties catalog properties
+   * @return an instance of a factory class
+   */
+  @SuppressWarnings("unchecked")
+  public static <T> T initFactory(Map<String, String> properties) {
```

Comment:
Nit: prefer simple names of possible, can we call this method just initialize?

Reply:
Actually, I think we can mostly follow the pattern in AwsClientFactories:

```
public static AwsClientFactory defaultFactory() {
    return AWS_CLIENT_FACTORY_DEFAULT;
  }

  public static AwsClientFactory from(Map<String, String> properties) {
    String factoryImpl =
        PropertyUtil.propertyAsString(
            properties, AwsProperties.CLIENT_FACTORY, DefaultAwsClientFactory.class.getName());
    return loadClientFactory(factoryImpl, properties);
  }
```

So we use a singleton if default, and don't need to use `PropertyUtil.propertyAsString(xxx, null)` which is a bit awkward.

Reply:
@jackye1995 when `S3FileIOProperties.CLIENT_FACTORY` is not set, we want to load the correct `AwsClientFactory` implementation defined in `AwsProperties.CLIENT_FACTORY` catalog property defined [here](https://github.com/apache/iceberg/blob/master/aws/src/main/java/org/apache/iceberg/aws/AwsClientFactories.java#L56). Thats why I set default value to `null`.

For ex, if someone sets `AwsProperties.CLIENT_FACTORY` to `AssumeRoleAwsClientFactory`. Then, we should support loading that factory impl class, right?